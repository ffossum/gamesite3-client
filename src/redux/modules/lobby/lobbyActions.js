/* @flow */
import type { GameDataState, GamesState } from '../games/gamesReducer';

export const CREATE_GAME_REQUEST = 'lobby/create game request';
export const GAME_CREATED = 'lobby/game created';
export const GAME_UPDATED = 'lobby/game updated';
export const ENTER_LOBBY = 'lobby/enter';
export const EXIT_LOBBY = 'lobby/exit';
export const REFRESH_LOBBY = 'lobby/refresh';

export type EnterLobbyAction = {
  type: typeof ENTER_LOBBY,
};
export function enterLobby(): EnterLobbyAction {
  return {
    type: ENTER_LOBBY,
  };
}

export type ExitLobbyAction = {
  type: typeof EXIT_LOBBY,
};
export function exitLobby(): ExitLobbyAction {
  return {
    type: EXIT_LOBBY,
  };
}

export type RefreshLobbyAction = {
  type: typeof REFRESH_LOBBY,
  payload: GamesState,
};
export function refreshLobby(lobby: GamesState): RefreshLobbyAction {
  return {
    type: REFRESH_LOBBY,
    payload: lobby,
  };
}

export type CreateGameRequestAction = {
  type: typeof CREATE_GAME_REQUEST,
  payload: {
    userId: string,
  },
};

export function createGameRequest(userId: string): CreateGameRequestAction {
  return {
    type: CREATE_GAME_REQUEST,
    payload: {
      userId,
    },
  };
}

export type GameCreatedAction = {
  type: typeof GAME_CREATED,
  payload: GameDataState,
};

export function gameCreated(gameData: GameDataState): GameCreatedAction {
  return {
    type: GAME_CREATED,
    payload: gameData,
  };
}

export type GameUpdatedAction = {
  type: typeof GAME_UPDATED,
  payload: $Shape<GameDataState>,
};
export function gameUpdated(
  partialGameData: $Shape<GameDataState>,
): GameUpdatedAction {
  return {
    type: GAME_UPDATED,
    payload: partialGameData,
  };
}

export type LobbyAction =
  | EnterLobbyAction
  | ExitLobbyAction
  | RefreshLobbyAction
  | GameUpdatedAction
  | CreateGameRequestAction
  | GameCreatedAction;
