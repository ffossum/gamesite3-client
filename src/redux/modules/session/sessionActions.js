/* @flow */
export const AUTHENTICATED_USER = 'user/authenticated';

export type AuthenticatedUserAction = {
  type: 'user/authenticated',
  payload: PublicUserData,
};
export function authenticatedUser(user: PublicUserData) {
  return {
    type: AUTHENTICATED_USER,
    payload: user,
  };
}
