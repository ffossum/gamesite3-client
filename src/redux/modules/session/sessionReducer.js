/* @flow */
import type { Action } from '../../actions';
import { REGISTRATION_SUCCESS } from '../registration/registrationActions';

export type SessionUser = {
  id: string,
  username: string,
};
export type SessionState = {
  user: ?SessionUser,
};
const initialState: SessionState = {
  user: null,
};

export default function sessionReducer(
  state: SessionState = initialState,
  action: Action
) {
  switch (action.type) {
    case REGISTRATION_SUCCESS:
      return {
        ...state,
        user: action.payload,
      };
    default:
      return state;
  }
}
