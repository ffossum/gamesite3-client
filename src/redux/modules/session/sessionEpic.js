/* @flow */
import { Observable } from 'rxjs';
import { AUTHENTICATED_USER } from './sessionActions';
import type { AuthenticatedUserAction } from './sessionActions';
import type { Store } from 'redux';
import type DeepstreamClient from '../../deepstreamClient';
import { playerJoined, playerLeft } from '../games/gameRoomActions';

type Dependencies = {
  deepstreamClient: DeepstreamClient,
};

export default function sessionEpic(
  action$: Observable<*>,
  store: Store<*, *>,
  { deepstreamClient }: Dependencies,
) {
  return action$
    .filter(action => action.type === AUTHENTICATED_USER)
    .switchMap((action: AuthenticatedUserAction) => {
      return deepstreamClient
        .subscribe('user:' + action.payload.id)
        .flatMap(data => {
          switch (data.t) {
            case 'player-joined': {
              const gameId = data.p.gid;
              const userId = data.p.uid;
              return [playerJoined(userId, gameId)];
            }
            case 'player-left': {
              const gameId = data.p.gid;
              const userId = data.p.uid;
              return [playerLeft(userId, gameId)];
            }
            default:
              return [];
          }
        });
    });
}
