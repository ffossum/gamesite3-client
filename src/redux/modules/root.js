/* @flow */
import { combineReducers } from 'redux';
import { combineEpics } from 'redux-observable';
import registration from './registration/registrationReducer';
import type { RegistrationState } from './registration/registrationReducer';
import registrationEpic from './registration/registrationEpic';
import session from './session/sessionReducer';
import type { SessionState } from './session/sessionReducer';
import sessionEpic from './session/sessionEpic';

import login from './login/loginReducer';
import type { LoginState } from './login/loginReducer';
import loginEpic from './login/loginEpic';

import chat from './chat/chatReducer';
import type { ChatState } from './chat/chatReducer';
import chatEpic from './chat/chatEpic';

import users from './users/usersReducer';
import userDataEpic from './users/userDataEpic';
import type { UsersState } from './users/usersReducer';

import lobby from './lobby/lobbyReducer';
import lobbyEpic from './lobby/lobbyEpic';
import type { LobbyState } from './lobby/lobbyReducer';

import games from './games/gamesReducer';
import type { GamesState } from './games/gamesReducer';
import gameRoomEpic from './games/gameRoomEpic';

export type State = {
  chat: ChatState,
  login: LoginState,
  registration: RegistrationState,
  session: SessionState,
  users: UsersState,
  lobby: LobbyState,
  games: GamesState,
};

export const rootReducer = combineReducers({
  chat,
  login,
  registration,
  session,
  users,
  lobby,
  games,
});

export const rootEpic = combineEpics(
  chatEpic,
  loginEpic,
  registrationEpic,
  sessionEpic,
  userDataEpic,
  lobbyEpic,
  gameRoomEpic,
);
