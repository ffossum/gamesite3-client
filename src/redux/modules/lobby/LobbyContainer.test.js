/* @flow */
/* eslint-env jest */
import {
  mapStateToProps,
  mapDispatchToProps,
  mergeProps,
} from './LobbyContainer';
import { refreshLobby, createGameRequest } from './lobbyActions';
import { rootReducer } from '../root';
import { authenticatedUser } from '../session/sessionActions';

describe('lobby container', () => {
  let state;
  let dispatch;

  beforeEach(() => {
    state = rootReducer(
      state,
      refreshLobby({
        game_id: {
          status: 'not_started',
          createdTime: '2017-06-25T08:33:29.565Z',
          host: 'other_user_id',
          players: ['other_user_id'],
          id: 'game_id',
        },
      }),
    );
    dispatch = jest.fn();
  });

  test('denormalizes games data', () => {
    const stateProps = mapStateToProps(state);
    const dispatchProps = mapDispatchToProps(dispatch);
    const props = mergeProps(stateProps, dispatchProps);

    expect(props.games).toMatchSnapshot();
  });

  test('does not dispatch create game action if user is not authenticated', () => {
    const stateProps = mapStateToProps(state);
    const dispatchProps = mapDispatchToProps(dispatch);
    const props = mergeProps(stateProps, dispatchProps);

    props.createGame();
    expect(dispatch).not.toHaveBeenCalled();
  });

  test('dispatches create game action if user is authenticated', () => {
    state = rootReducer(
      state,
      authenticatedUser({
        email: 'user@test.com',
        id: 'user_id',
        username: 'user_name',
      }),
    );

    const stateProps = mapStateToProps(state);
    const dispatchProps = mapDispatchToProps(dispatch);
    const props = mergeProps(stateProps, dispatchProps);

    props.createGame();

    expect(dispatch).toHaveBeenCalledWith(createGameRequest('user_id'));
  });
});
