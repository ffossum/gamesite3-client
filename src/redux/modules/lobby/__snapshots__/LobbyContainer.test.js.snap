// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lobby container denormalizes games data 1`] = `
Array [
  Object {
    "createdTime": "2017-06-25T08:33:29.565Z",
    "host": Object {
      "id": "other_user_id",
      "username": "",
    },
    "id": "game_id",
    "players": Array [
      Object {
        "id": "other_user_id",
        "username": "",
      },
    ],
    "status": "not_started",
  },
]
`;
